### DISCUSSION TITLE
improve increment_version_number and increment_build_number allowing target selection

### DISCUSSION THREAD

**Discussion by kikeenrique** [fastlane/fastlane#17515] [Ideas]
Hi! ✋

Currently, looking at the state of get_version_number, increment_version_number, get_build_number and increment_build_number.

At the moment:
**increment_version_number**, **get_build_number** and **increment_build_number**: 

- are based on agvtool 
- which do not allow target selection
- and [will surely fail as Xcode usually refers to plists using $(SRCROOT), which agvtool seems to ignore](https://stackoverflow.com/questions/49429944/cannot-find-srcroot-agvtool-increment-build-number-fastlane). 

**get_version_number**: 

- uses xcodeproj directly.
- allows target selection.

My suggestion is to improve increment_version_number, get_build_number and increment_build_number using xcodeproj and target plist.

Expected behaviour would be to mimic Xcode (as since Xcode 11):
* Modify xcodeproj vars CURRENT_PROJECT_VERSION and MARKETING_VERSION
* Modify target plist vars CFBundleShortVersionString with $(MARKETING_VERSION) and CFBundleVersion with $(CURRENT_PROJECT_VERSION)

It seems some others have worked on solutions before, but I think top tools like fastlane shall provide first class support for versioning control. 
To name a few:

- It exist a plugin named fastlane-plugin-versioning.

- [Moath Othman implemented a increment_build_number  and shared on Internet](https://moathothman.com/index.php/2019/11/30/xcode-increment-build-number-per-configuration/)

- Others may be directly using plist vars writing API to solve this (at me company at least)

So let me know how this proposal sounds to you!

Regards

**Comment by rogerluan** [comment]
Hey @kikeenrique! 👋 

I don't have much context into `increment_version_number` and `increment_build_number` internals, but your reasonings sound good to me. Feel free to open a pull request with your proposed changes, and tag me in there to review the code 👍 

Looking forward to seeing your contributions! 💪
